File: 2025/212. Word Search II/Solution.java
class TrieNode {
    String word;
    TrieNode[] children;
    public TrieNode() {
        this.word = null;
        this.children = new TrieNode[26];
    }
}
class Trie {
    TrieNode root;
    
    public Trie() {
        root = new TrieNode(); 
    }

    public void addWord(String word) {
        char[] arr = word.toCharArray();
        TrieNode cur = root;
        for (int i = 0; i < arr.length; i++) {
            int index = arr[i] - 'a';
            if (cur.children[index] == null) {
                cur.children[index] = new TrieNode();
            }
            cur = cur.children[index];
        }
        cur.word = word;
    }

    public String search(String word) {
        char[] arr = word.toCharArray();
        TrieNode cur = root;
        for (int i = 0; i < arr.length; i++) {
            int index = arr[i] - 'a';
            if (cur.children[index] == null) {
                return null;
            }
            cur = cur.children[index];
        }
        return word;
    }
}

class Solution {

    Set<String> set;
    Trie trie;
    boolean[][] visited;

    int[] dx = {0, 1, 0, -1};
    int[] dy = {1, 0, -1, 0};

    public List<String> findWords(char[][] board, String[] words) {
        int m = board.length, n = board[0].length;
        trie = new Trie();
        set = new HashSet();
        visited = new boolean[m][n];
        for (String w : words) {
            trie.addWord(w);
        }
        TrieNode root = trie.root;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                int index = board[i][j] - 'a';
                if (root.children[index] != null) {
                    search(board, i, j, root.children[index]);
                }
            }
        }
        return new ArrayList<>(set);
    }

    private void search(char[][] board, int i, int j, TrieNode node) {
        if (node.word != null) {
            set.add(node.word);
        }
        visited[i][j] = true;
        for (int k = 0; k < 4; k++) {
            int x = dx[k] + i, y = dy[k] + j;
            if (isWithinBoundary(board, x, y)) {
                int index = board[x][y] - 'a';
                if (node.children[index] != null)
                    search(board, x, y, node.children[index]);
            }
        }
        visited[i][j] = false;
    }

    private boolean isWithinBoundary(char[][] board, int i, int j) {
        if (i >= 0 && j >= 0 && i < board.length && j < board[0].length && !visited[i][j])
            return true;
        else
            return false;
    }
}

File: 2025/212. Word Search II/solution.py
class TrieNode:
    def __init__(self):
        self.children = defaultdict(TrieNode)
        self.is_word = False

class Trie:
    def __init__(self, words: List[str]):
        self.root = TrieNode()
        for w in words:
            self.insert(w)
    
    def insert(self, s: str):
        cur = self.root
        for ch in s:
            if ch not in cur.children:
                cur.children[ch] = TrieNode()
            cur = cur.children[ch]
        cur.is_word = True

    def search(self, s: str) -> bool:
        cur = self.root
        for ch in s:
            if ch not in cur.children:
                return False
            cur = cur.children[ch]
        return cur.is_word

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        trie = Trie(words)
        res, visited = set(), set()
        def find(board, cur, s, i, j):
            if cur and cur.is_word:
                res.add(s)
            visited.add((i, j))
            for (x, y) in ((0, 1), (0, -1), (1, 0), (-1, 0)):
                ii, jj = i + x, j + y
                if (ii < 0 or jj < 0 or 
                    ii >= len(board) or 
                    jj >= len(board[0]) or 
                    (ii, jj) in visited or 
                    board[ii][jj] not in cur.children):
                    continue
                find(board, cur.children[board[ii][jj]], s + board[ii][jj], ii, jj)
            visited.remove((i, j))
        
        for i in range(len(board)):
            for j in range(len(board[i])):
                cur = trie.root
                if board[i][j] in cur.children:
                    find(board, cur.children[board[i][j]], board[i][j], i, j)
        return list(res)